src/validation.cpp

```diff
@@ -1757,6 +1761,12 @@ static bool WriteUndoDataForBlock(const CBlockUndo& blockundo, BlockValidationSt
             return error("ConnectBlock(): FindUndoPos failed");
         if (!UndoWriteToDisk(blockundo, _pos, pindex->pprev->GetBlockHash(), chainparams.MessageStart()))
             return AbortNode(state, "Failed to write undo data");
+        // rev files are written in block height order, whereas blk files are written as blocks come in (often out of order)
+        // we want to flush the rev (undo) file once we've written the last block, which is indicated by the last height
+        // in the block file info as below
+        if (_pos.nFile < nLastBlockFile && static_cast<uint32_t>(pindex->nHeight) == vinfoBlockFile[_pos.nFile].nHeightLast) {
+            FlushUndoFile(_pos.nFile, true);
```

Comment:
Maybe I am missing something, but it looks to me that if blocks are being downloaded in-order, then the `_pos.nFile < nLastBlockFile` condition will never be true because those will always be equal?

Reply:
I am pretty sure I tested this but I can't convince myself that you're wrong by looking at the code so I will test this again.

Reply:
The way I see it is that `_pos.nFile < nLastBlockFile && ` can (should) be dropped and the second condition suffices - if we are writing the last (highest) block in the rev file, then also finalize it because we are not going to touch it again.

Reply:
Sorry for tardiness, work kept me busy, but I managed to verify that this is *not* working as it should. I will fix soon.

Reply:
No worries. I started at this problem long enough and finally it cracked.

Indeed `_pos.nFile < nLastBlockFile` is never true if blocks come in-height-order and this needs to be adjusted. However, just removing that condition, as I suggest above, is not ok because then the second condition will be true for every in-height-order write and this will cause excessive flushing and finalizing. See the proposed [patch](https://github.com/bitcoin/bitcoin/pull/17994#issuecomment-599031098) for a fully working solution (I hope).

Reply:
I've got something but testing takes time (I don't wanna have to wait days each time so I'm juggling 300 GB full node dirs around).